lib/linkerd2-exp-backoff/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+pub struct ExponentialBackoff {
+    pub min: Duration,
+    pub max: Duration,
+    pub jitter: f64,
+}
+
+/// A jittered exponential backoff stream.
+#[derive(Debug)]
+pub struct ExponentialBackoffStream {
+    backoff: ExponentialBackoff,
+    rng: SmallRng,
+    iterations: u32,
+    delay: Option<timer::Delay>,
+}
+
+#[derive(Clone, Debug)]
+pub struct InvalidBackoff(&'static str);
+
+impl ExponentialBackoff {
+    pub fn stream(&self) -> ExponentialBackoffStream {
+        ExponentialBackoffStream {
+            backoff: self.clone(),
+            rng: SmallRng::from_entropy(),
+            iterations: 0,
+            delay: None,
+        }
+    }
+}
+
+impl ExponentialBackoff {
+    pub fn new(min: Duration, max: Duration, jitter: f64) -> Result<Self, InvalidBackoff> {
+        if min > max {
+            return Err(InvalidBackoff("maximum must not be less than minimum"));
+        }
+        if max == Duration::from_millis(0) {
+            return Err(InvalidBackoff("maximum must be non-zero"));
+        }
+        if jitter < 0.0 {
+            return Err(InvalidBackoff("jitter must not be negative"));
+        }
+        Ok(ExponentialBackoff { min, max, jitter })
+    }
+
+    fn base(&self, failures: u32) -> Duration {
+        debug_assert!(failures >= 1, "failiures must be non-zero");
+        debug_assert!(
+            self.min <= self.max,
+            "maximum backoff must not be less than minimum backoff"
+        );
+        debug_assert!(
+            self.max > Duration::from_millis(0),
+            "Maximum backoff must be non-zero"
+        );
+        self.min.mul(2_u32.saturating_pow(failures)).min(self.max)
+    }
+
+    fn jitter<R: rand::Rng>(&self, base: Duration, rng: &mut R) -> Duration {
+        if self.jitter == 0.0 {
+            Duration::default()
+        } else {
+            let jitter_factor = rng.gen::<f64>();
+            debug_assert!(
+                jitter_factor > 0.0,
+                "rng returns values between 0.0 and 1.0"
+            );
+            let rand_jitter = jitter_factor * self.jitter;
+            let secs = (base.as_secs() as f64) * rand_jitter;
+            let nanos = (base.subsec_nanos() as f64) * rand_jitter;
+            let remaining = self.max - base;
+            Duration::new(secs as u64, nanos as u32).min(remaining)
+        }
+    }
+}
+
+impl Stream for ExponentialBackoffStream {
+    type Item = ();
+    type Error = timer::Error;
+
+    fn poll(&mut self) -> Poll<Option<()>, Self::Error> {
+        loop {
+            // If there's an active delay, wait until it's done and then
+            // update the state.
+            if let Some(delay) = self.delay.as_mut() {
+                try_ready!(delay.poll());
+
+                self.delay = None;
+                self.iterations += 1;
+                return Ok(Some(()).into());
+            }
+
+            let backoff = {
+                let base = self.backoff.base(self.iterations + 1);
```

Comment:
good catch

Reply:
I think it may be a little simpler to just pass in `self.iterations` without the ` + 1` and remove the `- 1` in `fn base`. I was wondering why that was there before seeing this conversation, but obviously not a blocker.